src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/ProjectConfigurationStateManager.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.ProjectSystem;
+using Microsoft.AspNetCore.Razor.Serialization;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+
+/// <summary>
+/// Updates project system service with new project info.
+/// </summary>
+/// <remarks>
+/// Used to figure out if the project system data is being added, updated or deleted
+/// and acts accordingly to update project system service.
+/// </remarks>
+internal class ProjectConfigurationStateManager
+{
+    private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+    private readonly IRazorProjectService _projectService;
+    private readonly ILogger _logger;
+
+    /// <summary>
+    /// A set of all projects known to this service.
+    /// </summary>
+    private readonly HashSet<ProjectKey> _projectSet;
+
+    /// <summary>
+    /// Used to throttle project system updates
+    /// </summary>
+    internal readonly Dictionary<ProjectKey, DelayedProjectInfo> ProjectInfoMap;
```

Comment:
Do we need to throttle? I'm not sure how expensive updating project info is, but from memory the RazorProjectService does some work to take the full info, and only updates the things that have changed, or do nothing if nothing has changed. Feels like any throttling that happens should be on the LSP client side of things. If something sends a `razor/projectInfo` message to the server, then immediately follows with a completion request, it would be odd for the tag helpers it just sent to not be in the completion list.

Reply:
The throttling was copied from ProjectConfigurationStateSynchronizer - I assumed it was needed :), though it's not necessarily true of course. It was added in ProjectConfigurationStateSynchronizer by

https://github.com/dotnet/razor/pull/1777/files#

The same PR also increased throttle delay to 3 seconds from 250 ms in https://github.com/dotnet/razor/blob/464030ffb65aba45250a1c200bc3528d5dd7a430/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorProjectInfoPublisher.cs#L61C18-L61C31

I'm not clear why the delay was needed both on the client and on the server, seems like the client side should've been enough - but it was done on both sides in the same PR.

I'm game to try removing the throttling here if you don't think it's necessary - what do you think?

Reply:
I think its definitely worth trying it without throttling, definitely on the receiving side anyway. The PR description says it was to "resolve the corrupted JSON problems", so those will definitely be a thing of the past. Also possibly explains the delay on both sides - as long as the server one is longer, its a fair assumption that the client has finished doing file writes. I could see having a small delay on the sending end being good, even with this change, but 3 seconds sounds extreme.

I think we probably need to come up with a metric we can add, to compare before and after. Maybe something like a counter for every time we compile a .razor file in the language server. If we add that, run through a simple scenario with the feature flag on and off, and compare numbers, we'll get an idea of whether we're causing a bunch more work to happen. A key thing to test would be with and without a file already being open from a previous run on VS. With no file open, the language server won't start and all of the initial noise of project loading will have no effect, since the new model simply won't send any data. Having a file pre-opened means we can measure the impact of that noise, too.

Reply:
I think maybe I understood why throttling is necessary on both sides. Even if we throttle on the client, the server has a queue of received requests, and if there are some long-ish requests on the server executing while several "project info" requests come from the client and are placed in the queue, then the "project info" requests from the server queue might end up getting dequeued and executed immediately one after another. So it seems like it is correct that we need to throttle on both sides:

1. On client so we throttle the requests we send to the server
2. On the server *after* the request are dequeued from the server-side request queue 

Does that sound reasonable or is my logic off somewhere?

Reply:
I'm on board with throttling on the client before sending, that sounds fine. For the server though, it seems like a it would be unnecessary. You raise a good point about overlapping requests though, but thats where setting `MutatesSolutionState` to `true` just completely solves the problem. We'll never have competing updates. BUT if thats a bad idea because of how long the requests take to process, then yes, a short delay is probably wise.

I don't feel like I have a good idea of what exactly happens with one of these updates. Does it just update state and produce new snapshots? Or do we kick off compilation and other expensive things? It's also not clear to me how often they happen, and actually have new data. Should just be new or deleted files or references I would have thought.

Maybe I'm overthinking though, and just using the `AsyncBatchingWorkQueue` with a few hundred millisecond delay is worth putting in, just in case.

Reply:
Seems simple enough to add batching now that I figured out how to use the batching queue, will add this tomorrow morning. Unless you think I really shouldn't :)

Reply:
Yeah,. the `AsyncBatchingWorkQueue` is the way to go.

Reply:
Implemented in https://github.com/dotnet/razor/pull/10182/commits/0586df0ebd84dfc375067a18dc6b9acd9aca1b93